testonly/hammer/hammer.go

```diff
@@ -186,51 +186,80 @@ func HitMap(cfg MapConfig) error {
 		}
 	}(ticker.C)
 
-	for count := uint64(1); count < cfg.Operations; count++ {
-		if err := s.retryOneOp(ctx); err != nil {
-			return err
+	var wg sync.WaitGroup
+	errs := make(chan error, cfg.NumCheckers+1)
+	done := make(chan struct{})
+	for i := 0; i < cfg.NumCheckers; i++ {
```

Comment:
Channels and their blocking behavior can be a bit tricky to reason about clearly. Would you mind putting this in the form of a channel pattern? 

eg. 
```go
func launchCheckers(n numCheckers) <- chan error {
    errs := make(chan error)
    go func() {
         wg := new(sync.WaitGroup)
         for i:=0; i< n; i++ {
             wg.Add(1)
             go func() {
                   defer wg.Done()
                   errs <- s.readChecker()
              }() 
          }
         wg.Wait()
         close(errs)
    }()
    return errs
}

Reply:
Your alternative seems harder to reason about to me, so I prefer to leave it as-is -- there's no blocking channel behaviour in the current code.

Reply:
Fair enough. The example might be the wrong approach. 

The thing that still seems a bit hairy is the multi-purpose nature of the  `errs` channel. Could we try simplifying it? 

* Some go routines only put errors in the `err` channel, others may put `nil`.
* Only the first error is used, but there's not really a guarantee about the order of items in the channel.   (If nothing produces an error, then we know that there will only be one `nil`, but this seems fragile) 

Consider separating:

* Consuming and logging all errors. 
* Determining whether the main routine completed successfully.
* ~Waiting for completion.~ (Already done with the wait group)